pkg/cloudprovider/aws/suite_test.go

```diff
@@ -538,6 +579,112 @@ var _ = Describe("Allocation", func() {
 				}
 			})
 		})
+		Context("MetadataOptions", func() {
+			It("should allow missing values", func() {
+				provider, err := ProviderFromProvisioner(provisioner)
```

Comment:
Instead of this, I would just use the scoped variable `provider`, mutate it, and then call ProvisionerWithProvider. Other tests follow this pattern.

Reply:
I think there is actually a bug with the other tests. Currently the defaulting of some of the provider fields are handled in the `BeforeEach`. Those defaults are only persisted into a copy of the provider that is deserialized into the provisioned. As a result when we do `ProviderFromProvisioner` again some of the defaulted values are reverted back to being unset. This results in validation errors that are unrelated to the specific test case. I think this is unnoticed because all existing test cases that mutate the provider expect an unqualified validation error.

For example, if I have the following tests, the second one fails due to `missing field(s): spec.provider.securityGroupSelector, spec.provider.subnetSelector`:
```
		FIt("should validate", func() {
			Expect(provisioner.Validate(ctx)).To(Succeed())
		})
		FIt("should validate", func() {
			Expect(ProvisionerWithProvider(provisioner, provider).Validate(ctx)).To(Succeed())
		})
```

Because defaulting the provider fields is encapsulated inside the `(p *Provisioner) SetDefaults(ctx context.Context)`, I'm not sure of a cleaner way to default the provider fields. I'm open to other ideas though.

Reply:
Should we call SetDefaults() inside of ProvisionerWithProvider to enforce this? That will ensure the defaults are set for the test, but won't override anything if we want to set something else. WDYT?

Reply:
I had considered that originally, but some of the existing tests are trying to `nil` out fields to ensure they'd be considered invalid. If we try to default them again we'll override the intention of the test and cause the test to fail.

Reply:
@ellistarn I was just circling back on this. Let me know if you want me to do anything further here. I think that without expanding the API of the `v1alpha1.AWS` struct the only way to obtain a provider that's been defaulted to a valid configuration is to pull it out of a provisioner that's been defaulted. I can certainly adjust the approach here and bind new functions to the `v1alpha1.AWS` struct but I'm not sure if the intent was to keep that API surface-area minimal or not.

If the current approach is suitable given existing constraints, it would be cool if we can get this integrated. 👍

Reply:
I'll bias for merging since it's test code. I'll clean it up later if it bothers me. Thanks for your contribution!